### PR TITLE
Optimizer for aggregations with distinct mask

### PR DESCRIPTION
**Pull request by stagraqubole**
Optimizes queries which have aggregations on both distinct mask and other symbols

E.g.: 

```
presto:default> explain select o_custkey, COUNT(o_shippriority), sum(DISTINCT o_orderkey) FROM orders group by o_custkey;
                                                                                                                                  Query Plan                                                                                                                              
--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
 - Output[o_custkey, _col1, _col2] => [o_custkey:integer, count:bigint, sum:bigint]                                                                                                                                                                                       
         _col1 := count                                                                                                                                                                                                                                                   
         _col2 := sum                                                                                                                                                                                                                                                     
     - RemoteExchange[GATHER] => o_custkey:integer, count:bigint, sum:bigint                                                                                                                                                                                              
         - Project => [o_custkey:integer, count:bigint, sum:bigint]                                                                                                                                                                                                       
             - Aggregate(FINAL)[o_custkey] => [o_custkey:integer, $hashvalue:bigint, count:bigint, sum:bigint]                                                                                                                                                            
                     count := "arbitrary"("arbitrary")                                                                                                                                                                                                                    
                     sum := "sum"("sum_18")                                                                                                                                                                                                                               
                 - RemoteExchange[REPARTITION] => o_custkey:integer, arbitrary:bigint, sum_18:bigint, $hashvalue:bigint                                                                                                                                                   
                     - Aggregate(PARTIAL)[o_custkey] => [o_custkey:integer, $hashvalue_21:bigint, arbitrary:bigint, sum_18:bigint]                                                                                                                                        
                             arbitrary := "arbitrary"("expr_16")                                                                                                                                                                                                          
                             sum_18 := "sum"("expr_15")                                                                                                                                                                                                                   
                         - Project => [o_custkey:integer, $hashvalue_21:bigint, expr_15:bigint, expr_16:bigint]                                                                                                                                                           
                                 $hashvalue_21 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("o_custkey"), 0))                                                                                                                                             
                                 expr_15 := (CASE WHEN ("group" = CAST(1 AS BIGINT)) THEN "o_orderkey_2" END)                                                                                                                                                             
                                 expr_16 := (CASE WHEN ("group" = CAST(0 AS BIGINT)) THEN "count_14" END)                                                                                                                                                                 
                             - Aggregate(FINAL)[o_custkey, o_orderkey_2, group] => [o_custkey:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_19:bigint, count_14:bigint]                                                                                          
                                     count_14 := "count"("count_17")                                                                                                                                                                                                      
                                 - RemoteExchange[REPARTITION] => o_custkey:integer, o_orderkey_2:bigint, group:bigint, count_17:bigint, $hashvalue_19:bigint                                                                                                             
                                     - Aggregate(PARTIAL)[o_custkey, o_orderkey_2, group] => [o_custkey:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_20:bigint, count_17:bigint]                                                                                
                                             count_17 := "count"("o_shippriority")                                                                                                                                                                                        
                                         - Project => [o_custkey:integer, o_shippriority:integer, o_orderkey_2:bigint, group:bigint, $hashvalue_20:bigint]                                                                                                                
                                                 $hashvalue_20 := "combine_hash"("combine_hash"("combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("o_custkey"), 0)), COALESCE("$operator$hash_code"("o_orderkey_2"), 0)), COALESCE("$operator$hash_code"("group"),
                                             - GroupId[[o_custkey, o_shippriority], [o_custkey, o_orderkey_2]] => [o_custkey:integer, o_shippriority:integer, o_orderkey_2:bigint, group:bigint]                                                                          
                                                 - Project => [o_shippriority:integer, o_orderkey_2:bigint, o_custkey:integer]                                                                                                                                            
                                                         o_orderkey_2 := CAST("o_orderkey" AS BIGINT)                                                                                                                                                                     
                                                     - TableScan[hive:hive:default:orders, originalConstraint = true] => [o_orderkey:integer, o_custkey:integer, o_shippriority:integer]                                                                                  
                                                             LAYOUT: hive                                                                                                                                                                                                 
                                                             o_orderkey := HiveColumnHandle{clientId=hive, name=o_orderkey, hiveType=int, hiveColumnIndex=0, columnType=REGULAR}                                                                                          
                                                             o_custkey := HiveColumnHandle{clientId=hive, name=o_custkey, hiveType=int, hiveColumnIndex=1, columnType=REGULAR}                                                                                            
                                                             o_shippriority := HiveColumnHandle{clientId=hive, name=o_shippriority, hiveType=int, hiveColumnIndex=7, columnType=REGULAR}     
```

### CODE REVIEW

**Comment by stagraqubole** [comment]
Checked performance with this query:
```
SELECT ss_store_sk, 
       Sum(ss_quantity)               AS s, 
       Count(DISTINCT ss_customer_sk) AS d 
FROM   store_sales_orc_100
GROUP  BY ss_store_sk
```

With optimization, it works 2x-2.5x faster.

This optimizer can be extended to handle multiple distinct aggregations on different masks as well.

**Comment by maciejgrzybek** [comment]
@stagraqubole thank you for this contribution!
Can you add some DSL testing (https://github.com/prestodb/presto/pull/4404)? Example of use: https://github.com/prestodb/presto/pull/5714/commits/27de6dbe9af62b5e21166822bc1e7e76534a6057

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 1]
```diff
@@ -0,0 +1,436 @@
...
+ *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group FROM
+ *        SELECT a1, a2,..., an, b1, b2, ... ,bn, c FROM Table GROUP BY GROUPING SETS ((a1, a2,..., an, b1, b2, ... ,bn), (a1, a2,..., an, c))
+ *      GROUP BY a1, a2,..., an, c, group
+ *  GROUP BY a1, a2,..., an
+ */
+public class DistinctOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private static final String HASH_CODE = FunctionRegistry.mangleOperatorName("HASH_CODE");
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: Not used

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 2]
```diff
@@ -0,0 +1,436 @@
...
+                    Optional.<Symbol>empty());
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
+        {
+            Optional<AggregateInfo> aggregateInfo = context.get();
+            // presence of aggregateInfo => mask also present
+            if (aggregateInfo.isPresent() && aggregateInfo.get().getMask().get().equals(node.getMarkerSymbol())) {
+                PlanNode source = node.getSource();
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: Call rewrite on source?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+ *      GROUP BY a1, a2,..., an, c, group
+ *  GROUP BY a1, a2,..., an
+ */
+public class DistinctOptimizer
+        implements PlanOptimizer
+{
+    private final Metadata metadata;
+    private static final String HASH_CODE = FunctionRegistry.mangleOperatorName("HASH_CODE");
+
+    public DistinctOptimizer(Metadata metadata)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: Maybe give this class more meanigful name. `ExtractDistinctAggregationOptimizer`?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+                    Collections.emptyMap(),
+                    node.getGroupingSets(),
+                    node.getStep(),
+                    node.getSampleWeight(),
+                    node.getConfidence(),
+                    Optional.<Symbol>empty());
+        }
+
+        @Override
+        public PlanNode visitMarkDistinct(MarkDistinctNode node, RewriteContext<Optional<AggregateInfo>> context)
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: If you code extract parts of this method to other methods - it could make it more readable. It is very long and deep.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+                Symbol groupSymbol = symbolAllocator.newSymbol("group", BigintType.BIGINT); // g
+
+                List<List<Symbol>> groups = new ArrayList<>();
+                // g0 = {allGBY_Symbols + allNonDistinctAggregateSymbols}
+                // g1 = {allGBY_Symbols + Distinct Symbol}
+                // symbols present in Group_i will be set, rest will be Null
+
+                //g0
+                if (!nonDistinctAggregateSymbols.isEmpty()) {
```
Comment (losipiuk) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: Is that possible provided the check `(masks.size() != 1 || node.getMasks().size() == node.getAggregations().size())` you do before.

**Comment by losipiuk** [comment]
Could you explain why this optimization improves performance?
Maybe this is good as short term solution but actually we should improve performance of execution part which is responsible for distinct aggregations.

**Comment by stagraqubole** [comment]
@maciejgrzybek @losipiuk thanks for the comments, I will resolve them.

Regarding performance, improvement is because of the lesser data transferred. With the MarkDistinct plan, we transfer all data from SOURCE stage to next, with this optimization we are are partially grouping by the distinct symbol at SOURCE stage and sending data.

File: presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java [hunk 6]
```diff
@@ -61,6 +61,7 @@
     public static final String SPLIT_CONCURRENCY_ADJUSTMENT_INTERVAL = "split_concurrency_adjustment_interval";
     public static final String OPTIMIZE_METADATA_QUERIES = "optimize_metadata_queries";
     public static final String QUERY_PRIORITY = "query_priority";
+    public static final String OPTIMIZE_DISTINCT_AGGREGATIONS = "optimize_distinct_aggregations";
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/SystemSessionProperties.java]: Please extract adding session property to separate commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,436 @@
...
+    public DistinctOptimizer(Metadata metadata)
+    {
+        this.metadata = metadata;
+    }
+
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, Map<Symbol, Type> types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator)
+    {
+        if (SystemSessionProperties.isOptimizeDistinctAggregationEnabled(session)) {
+            return SimplePlanRewriter.rewriteWith(new Optimizer(idAllocator, symbolAllocator, metadata),
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: static import

**Comment by kokosing** [comment]
I have just skimmed, I will look more closely tomorrow.

**Comment by kokosing** [comment]
@stagraqubole back to you

**Comment by sopel39** [comment]
This is an interesting optimization. Another idea would be rewrite `MarkDistinct` into `AggregationOperator` that would compute original `F1,...Fm`. It would also compute `c` distinct values partially at `SOURCE` stage and apply `F` on distinct `c` in final aggregation stage.
Such approach would reduce processing overhead and inter node pages copying between operators (e.g: overall plan would be simpler).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java (hunk 5 above)
Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/DistinctOptimizer.java]: You are right, wont happen right now. This is there because I was thinking of extending this to multiple distinct aggregations. Will remove it for now.

**Comment by stagraqubole** [comment]
@sopel39 This in a way does that as it groups by mask symbol in first aggregation.

I have made changes as per review comments:
- Added DSL tests
- Changed optimizer name and refactored code a bit
- Added separate commit for session property

@kokosing @losipiuk @maciejgrzybek ready for another round of review

**Comment by martint** [comment]
@erichwang, can you review?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+                this.mask = Optional.of(mask);
+            }
+
+            this.aggregations = ImmutableMap.copyOf(aggregations);
+            this.functions = ImmutableMap.copyOf(functions);
+        }
+
+        public List<Symbol> getOriginalNonDistinctAggregateArgs()
+        {
+            Set<Symbol> arguments = new HashSet<Symbol>();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: You can write it like this, instead:

```java
return aggregations.values().stream()
        .filter(function -> !function.isDistinct())
        .flatMap(function -> function.getArguments().stream())
        .distinct()
        .map(Symbol::from)
        .collect(Collectors.toList());
```

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+            aggregations.values()
+                    .stream()
+                    .filter(functionCall -> !functionCall.isDistinct())
+                    .forEach(functionCall -> functionCall.getArguments().forEach(argument -> arguments.add(Symbol.from(argument))));
+            return arguments.stream().collect(Collectors.toList());
+        }
+
+        public List<Symbol> getOriginalDistinctAggregateArgs()
+        {
+            Set<Symbol> arguments = new HashSet<Symbol>();
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Same as above

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+            aggregations.values()
+                    .stream()
+                    .filter(functionCall -> functionCall.isDistinct())
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: FunctionCall::isDistinct

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+                else {
+                    // Aggregations on non-distinct are already done by new node, just extract the non-null value
+                    Symbol argument = aggregateInfo.getNewNonDistinctAggregateSymbols().get(entry.getKey());
+                    QualifiedName functionName = QualifiedName.of("arbitrary");
+                    aggregations.put(entry.getKey(), new FunctionCall(functionName,
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(argument.toSymbolReference())));
+                    functions.put(entry.getKey(),
+                            metadata.getFunctionRegistry().resolveFunction(functionName, ImmutableList.of(symbolAllocator.getTypes().get(argument).getTypeSignature()), false));
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: This is a little hard to read. I'd either introduce a variable for the parameter types or the function or format it across multiple lines.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 12]
```diff
@@ -0,0 +1,482 @@
...
+                }
+            }
+
+            return new AggregationNode(idAllocator.getNextId(),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: First argument goes on the next line.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+                    SINGLE,
+                    Optional.empty(),
+                    1.0,
+                    originalNode.getHashSymbol());
+        }
+
+        // creates if clause specific to use case here, default value always null
+        private SearchedCaseExpression createIfExpression(Expression left, Expression right, ComparisonExpression.Type type, Expression result)
+        {
+            return new SearchedCaseExpression(ImmutableList.of(createWhenClause(left, right, type, result)),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: You can use IfExpression directly

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 14]
```diff
@@ -0,0 +1,482 @@
...
+            //          SELECT a1, a2,..., an, arbitrary(f1 if group = 0 else null),...., arbitrary(fm if group = 0 else null), F(c if group = 1 else null)
+            ImmutableMap.Builder<Symbol, FunctionCall> aggregations = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Signature> functions = ImmutableMap.builder();
+            for (Map.Entry<Symbol, FunctionCall> entry : node.getAggregations().entrySet()) {
+                FunctionCall functionCall = entry.getValue();
+                if (entry.getValue().isDistinct()) {
+                    aggregations.put(entry.getKey(), new FunctionCall(functionCall.getName(),
+                            functionCall.getWindow(),
+                            false,
+                            ImmutableList.of(Iterables.getOnlyElement(aggregateInfo.getNewDistinctAggregateSymbols().values()).toSymbolReference())));
```
Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Will change it, this was left as map  because I originally thought of writing this for multiple distinct aggregations.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 15]
```diff
@@ -0,0 +1,482 @@
...
+
+        // Filled on the way back, these are the symbols corresponding to their distinct or non-distinct original symbols
+        private Map<Symbol, Symbol> newNonDistinctAggregateSymbols;
+        private Map<Symbol, Symbol> newDistinctAggregateSymbols;
+
+        public AggregateInfo(List<Symbol> groupBySymbols, Symbol mask, Map<Symbol, FunctionCall> aggregations, Map<Symbol, Signature> functions)
+        {
+            this.groupBySymbols = ImmutableList.copyOf(groupBySymbols);
+
+            if (mask != null) {
```
Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: right, this is unnecessary, will remove it.

**Comment by stagraqubole** [comment]
Taken care of review comments @martint

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 16]
```diff
@@ -0,0 +1,481 @@
...
+                )
+        {
+            ImmutableMap.Builder<Symbol, Expression> outputSymbols = ImmutableMap.builder();
+            ImmutableMap.Builder<Symbol, Symbol> outputNonDistinctAggregateSymbols = ImmutableMap.builder();
+            for (Symbol symbol : source.getOutputSymbols()) {
+                if (distinctSymbol.equals(symbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol("expr", symbolAllocator.getTypes().get(symbol));
+                    aggregateInfo.setNewDistinctAggregateSymbol(newSymbol);
+
+                    Expression expression = createIfExpression(groupSymbol.toSymbolReference(),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: This is probably ok for now, but it should eventually be replaced with `GROUPING` when it's available: https://github.com/prestodb/presto/pull/5465

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: First argument goes on next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 17]
```diff
@@ -0,0 +1,481 @@
...
+                            new Cast(new LongLiteral("1"), "bigint"),
+                            ComparisonExpression.Type.EQUAL,
+                            symbol.toSymbolReference(),
+                            symbolAllocator.getTypes().get(symbol));
+                    outputSymbols.put(newSymbol, expression);
+                }
+                else if (aggregationOuputSymbolsMap.containsKey(symbol)) {
+                    Symbol newSymbol = symbolAllocator.newSymbol("expr", symbolAllocator.getTypes().get(symbol));
+                    outputNonDistinctAggregateSymbols.put(aggregationOuputSymbolsMap.get(symbol), newSymbol); // key of this map is key of an aggregation in AggrNode above, it will now aggregate on this Map's value
+                    Expression expression = createIfExpression(groupSymbol.toSymbolReference(),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: First argument goes on next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 18]
```diff
@@ -0,0 +1,481 @@
...
+                }
+            }
+
+            // add null assignment for mask
+            // unused mask will be removed by PruneUnreferencedOutputs
+            outputSymbols.put(aggregateInfo.getMask().get(), new NullLiteral());
+
+            aggregateInfo.setNewNonDistinctAggregateSymbols(outputNonDistinctAggregateSymbols.build());
+
+            return new ProjectNode(idAllocator.getNextId(),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Place all arguments on the same line. It's not too long in this case.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 19]
```diff
@@ -0,0 +1,481 @@
...
+                    source,
+                    outputSymbols.build());
+        }
+
+        private GroupIdNode createGroupIdNode(List<Symbol> groupBySymbols,
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: First argument goes on next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 20]
```diff
@@ -0,0 +1,481 @@
...
+         * This method returns a new Aggregation node which has aggregations on non-distinct symbols from original plan. Generates
+         *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group
+         * part in the optimized plan mentioned above
+         *
+         * It also populates the mappings of new function's output symbol to corresponding old function's output symbol, e.g.
+         *     { f1 -> F1, f2 -> F2, ... }
+         * The new AggregateNode aggregates on the symbols that original AggregationNode aggregated on
+         * Original one will now aggregate on the output symbols of this new node
+         */
+        private AggregationNode createNonDistinctAggregation(AggregateInfo aggregateInfo,
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: First argument goes on next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 21]
```diff
@@ -0,0 +1,481 @@
...
+import static java.util.Objects.requireNonNull;
+
+/*
+ * This optimizer convert query of form:
+ *  SELECT a1, a2,..., an, F1(b1), F2(b2), F3(b3), ...., Fm(bm), F(distinct c) FROM Table GROUP BY a1, a2, ..., an
+ *
+ *  INTO
+ *  SELECT a1, a2,..., an, arbitrary(f1 if group = 0 else null),...., arbitrary(fm if group = 0 else null), F(c if group = 1 else null) FROM
+ *      SELECT a1, a2,..., an, F1(b1) as f1, F2(b2) as f2,...., Fm(bm) as fm, c, group FROM
+ *        SELECT a1, a2,..., an, b1, b2, ... ,bn, c FROM Table GROUP BY GROUPING SETS ((a1, a2,..., an, b1, b2, ... ,bn), (a1, a2,..., an, c))
```
Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Yes, it can happen but only if there are some aggregate functions which can work on HyperLogLog. There are no such functions today, not sure if aggregate functions on HyperLogLog are useful. Should I just add the check to bail out for non-equatable types to be on safer side?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java [hunk 22]
```diff
@@ -0,0 +1,481 @@
...
+                    duplicatedDistinctSymbol = newSymbol;
+
+                    // Add project node to get stream for new symbol
+                    ImmutableMap.Builder builder = new ImmutableMap.Builder();
+                    for (Symbol symbol : node.getSource().getOutputSymbols()) {
+                        builder.put(symbol, symbol.toSymbolReference());
+                    }
+                    // Cast to avoid UnaliasSymbolReference from removing this projection
+                    // TODO: find better way to prevent UnaliaSymbolRefrence optimizing out the project
+                    builder.put(newSymbol, new Cast(distinctSymbol.toSymbolReference(),
```
Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Let it be here? Or wait for #6379 to be fixed?

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: It's a correctness issue, which we treat with very high priority. We'll fix it shortly.

Comment (stagraqubole) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/ExtractDistinctAggregationOptimizer.java]: Ok, I will update the PR once it is fixed

**Comment by yuananf** [comment]
@stagraqubole will this kind of query be optimized?
```
presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_sold_time_sk), count(distinct ss_item_sk) from store_sales;
 _col0 | _col1  
-------+--------
 46800 | 300000 
(1 row)

Query 20161024_075438_00027_4gzht, FINISHED, 10 nodes
Splits: 1,905 total, 1,905 done (100.00%)
2:14 [2.88B rows, 9.84GB] [21.5M rows/s, 75.1MB/s]
```
It finished in over 2 minutes, but the following queries finished in less than 10 seconds.

```
presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_sold_time_sk) from store_sales;
 _col0 
-------
 46800 
(1 row)

Query 20161024_075416_00025_4gzht, FINISHED, 10 nodes
Splits: 1,895 total, 1,895 done (100.00%)
0:03 [2.88B rows, 2.76GB] [827M rows/s, 813MB/s]

presto:tpcds_bin_partitioned_orc_1000> select count(distinct ss_item_sk) from store_sales;
 _col0  
--------
 300000 
(1 row)

Query 20161024_075424_00026_4gzht, FINISHED, 10 nodes
Splits: 1,895 total, 1,895 done (100.00%)
0:06 [2.88B rows, 8.5GB] [449M rows/s, 1.33GB/s]

```

**Comment by stagraqubole** [comment]
@yuananf current PR will not optimize the query you mentioned. Right now, queries with single distinct mask and multiple aggregations will be optimized.
The plan is to extend these changes to handle multiple distinct masks as well. I will be working on it once this PR is approved. With those changes, the query you mentioned will also be optimized

**Comment by yuananf** [comment]
@stagraqubole thanks, that would be great.

**Comment by stagraqubole** [comment]
@martint took care of review comments, back to you

**Comment by martint** [comment]
Merged (with a few minor changes). Thanks!

BTW, I renamed the optimizer, config and session properties.
